src/upgrade.rs

```diff
@@ -0,0 +1,283 @@
+//! Update toolchain crates properly.
+
+use anyhow::Result;
+use clap::{Parser, ValueEnum};
+use semver::Version;
+use std::mem;
+use std::path::PathBuf;
+use toml_edit::{DocumentMut, InlineTable, Value};
+use xshell::{cmd, Shell};
+
+/// Update toolchain crates properly.
+#[derive(Parser)]
+pub struct Args {
+    /// Name of toolchain dependency (group) to update.
+    dep: DepName,
+
+    #[command(flatten)]
+    spec: Spec,
+
+    /// Do not edit any files, just inform what would be done.
+    #[arg(long, default_value_t = false)]
+    dry_run: bool,
+}
+
+#[derive(ValueEnum, Copy, Clone, Debug)]
+enum DepName {
+    Cairo,
+    #[value(name = "cairols")]
+    CairoLS,
+}
+
+#[derive(clap::Args, Clone)]
+#[group(required = true, multiple = true)]
+struct Spec {
+    /// Source the dependency from crates.io and use a specific version.
+    version: Option<Version>,
+
+    /// Source the dependency from the GitHub repository and use a specific commit/ref.
+    #[arg(short, long, conflicts_with = "branch")]
+    rev: Option<String>,
+
+    /// Source the dependency from the GitHub repository and use a specific branch.
+    #[arg(short, long)]
+    branch: Option<String>,
+
+    /// Source the dependency from a local filesystem.
+    ///
+    /// This is useful for local development, but avoid commiting this to the repository.
+    #[arg(short, long, conflicts_with_all = ["rev", "branch"])]
+    path: Option<PathBuf>,
+}
+
+pub fn main(args: Args) -> Result<()> {
+    let sh = Shell::new()?;
+
+    let mut cargo_toml = sh.read_file("Cargo.toml")?.parse::<DocumentMut>()?;
+
+    edit_dependencies(&mut cargo_toml, "dependencies", &args);
+    edit_dependencies(&mut cargo_toml, "dev-dependencies", &args);
```

Comment:
Does it work for workspaces without package in root? 🤔 (i.e. can this be defined when there is no package?)

Reply:
I don't expect such cases to happen anytime soon so I don't cover it

Reply:
Scarb has `[workspace.dependencies]` which is covered here as a third branch.